BitFaster.Caching/Lru/AfterReadStopwatchPolicy.cs

```diff
@@ -0,0 +1,126 @@
+﻿using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace BitFaster.Caching.Lru
+{
+#if !NETCOREAPP3_0_OR_GREATER
+    /// <summary>
+    /// Implement an expire after access policy.
+    /// </summary>    
+    /// <remarks>
+    /// This class measures time using Stopwatch.GetTimestamp() with a resolution of ~1us.
+    /// </remarks>
+    public readonly struct AfterAccessLongTicksPolicy<K, V> : IItemPolicy<K, V, LongTickCountLruItem<K, V>>
+    {
+        private readonly long timeToLive;
+        private readonly Time time;
+
+        /// <summary>
+        /// Initializes a new instance of the TLruLongTicksPolicy class with the specified time to live.
+        /// </summary>
+        /// <param name="timeToLive">The time to live.</param>
+        public AfterAccessLongTicksPolicy(TimeSpan timeToLive)
+        {
+            this.timeToLive = StopwatchTickConverter.ToTicks(timeToLive);
+            this.time = new Time();
+        }
+
+        ///<inheritdoc/>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public LongTickCountLruItem<K, V> CreateItem(K key, V value)
+        {
+            return new LongTickCountLruItem<K, V>(key, value, Stopwatch.GetTimestamp());
+        }
+
+        ///<inheritdoc/>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Touch(LongTickCountLruItem<K, V> item)
+        {
+            item.TickCount = this.time.Last;
+            item.WasAccessed = true;
+        }
+
+        ///<inheritdoc/>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Update(LongTickCountLruItem<K, V> item)
+        {
+            item.TickCount = Stopwatch.GetTimestamp();
+        }
+
+        ///<inheritdoc/>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool ShouldDiscard(LongTickCountLruItem<K, V> item)
+        {
+            this.time.Last = Stopwatch.GetTimestamp();
+            if (this.time.Last - item.TickCount > this.timeToLive)
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        ///<inheritdoc/>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool CanDiscard()
+        {
+            return true;
+        }
+
+        ///<inheritdoc/>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public ItemDestination RouteHot(LongTickCountLruItem<K, V> item)
+        {
+            if (this.ShouldDiscard(item))
```

Comment:
There is a possible optimization here (which also applies to TLRu): all of the route methods call into ShouldDiscard, which evaluates the time.

When warmed up, RouteHot is always called first (because we always try to dequeue from hot). During warm up, this may not happen.

Reply:
The RouteHot trick could also be applied to TrimExpired routine, which will also make many dupe time calls.